### ISSUE TITLE
Parser cannot parse set operations combining SELECT * with SELECT <column list>

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#10699]
The following valid SQL query fails to parse:

```sql
select *
from t1
union all
select a
from t2
```

The error is: 

```
Select list must contain 0 columns. Got: 1: [5:1] select *
from t1
union all
select a
[*]from t2
```

We must not check for column size correctness in set operations when we don't know the row size of one of the subqueries

**Comment by lukaseder** [comment]
The degree check is generally implemented wrong / too early.

We're currently checking for the correct degree of a `SELECT` subquery in a set operation already when the select list has been parsed. But at this point, we can't resolve asterisks yet, so we don't know the degree, which is one of the reasons why this special case here doesn't parse.

We can check things once we have parsed the entire set operation subquery, prior to assembling the subqueries in a combined query.

**Comment by lukaseder** [comment]
I'll keep the old degree check for now to double-check. Removing it breaks a few test assertions, e.g. when this isn't expected to parse: `(a, b) = any (select 1)`

**Comment by lukaseder** [comment]
Fixed in 3.15.0 and 3.14.4 (#10973)
